cpp/src/arrow/type.h

```diff
@@ -250,15 +251,55 @@ struct ListType : public DataType {
   static char const* name() { return "list"; }
 
   std::string ToString() const override;
+
+ protected:
+  // Constructor for classes that are implemented as List Arrays.
+  ListType(const std::shared_ptr<DataType>& value_type, Type::type logical_type)
+      : DataType(logical_type) {
+    // TODO ARROW-187 this can technically fail, make a constructor method ?
+    children_ = {std::make_shared<Field>("item", value_type)};
+  }
+};
+
+// BinaryType type is reprsents lists of 1-byte values.
+struct BinaryType : public ListType {
+  BinaryType() : BinaryType(Type::BINARY) {}
+  static char const* name() { return "binary"; }
+  std::string ToString() const override;
+
+ protected:
+  // Allow subclasses to change the logical type.
+  explicit BinaryType(Type::type logical_type)
+      : ListType(std::shared_ptr<DataType>(new UInt8Type()), logical_type) {}
 };
 
-// String is a logical type consisting of a physical list of 1-byte values
-struct StringType : public DataType {
-  StringType();
+// UTF encoded strings
+struct StringType : public BinaryType {
+  StringType() : BinaryType(Type::STRING) {}
 
   static char const* name() { return "string"; }
 
   std::string ToString() const override;
+
+ protected:
+  explicit StringType(Type::type logical_type) : BinaryType(logical_type) {}
+};
+
+struct CharType : public StringType {
+  int size;
+
+  explicit CharType(int size) : StringType(Type::CHAR), size(size) {}
+
+  CharType(const CharType& other) : CharType(other.size) {}
+  bool Equals(const DataType* other) const override {
+    return StringType::Equals(other) &&
+           (size == static_cast<const CharType*>(other)->size);
+  }
+  static char const* name() { return "char"; }
+  std::string ToString() const override;
+  // TODO(emkornfield) This is byte size not character size, which generally isn't how
+  // DBMSes work.
+  int value_size() const override { return size; }
```

Comment:
http://docs.oracle.com/cd/B19306_01/server.102/b14225/ch7progrunicode.htm (maximum number of characters) x (maximum number of bytes for each character)

Postgres doesn't seem to distinguish types (https://www.postgresql.org/docs/9.0/static/datatype-character.html) i.e. it seems to always stores a length per entry

Mysql (http://dev.mysql.com/doc/refman/5.7/en/charset-unicode-utf8.html) To save space with UTF-8, use VARCHAR instead of CHAR. Otherwise, MySQL must reserve three bytes for each character in a CHAR CHARACTER SET utf8 column because that is the maximum possible character length. For example, MySQL must reserve 30 bytes for a CHAR(10) CHARACTER SET utf8 column.

MS SQL https://msdn.microsoft.com/en-us/library/ms186939.aspx "n defines the string length and must be a value from 1 through 4,000. The storage size is two times n bytes. When the collation code page uses double-byte characters, the storage size is still n bytes."